pkg/applicationserver/applicationserver.go

```diff
@@ -588,6 +592,202 @@ func (as *ApplicationServer) handleJoinAccept(ctx context.Context, ids ttnpb.End
 	return nil
 }
 
+// downlinkQueuesUpdater creates the pending and current session downlink queues.
+// The updater generally mutates the end device's sessions.
+type downlinkQueuesUpdater func(context.Context, *ttnpb.EndDevice) ([]*ttnpb.ApplicationDownlink, []*ttnpb.ApplicationDownlink, error)
+
+// changeDownlinkQueue updates the downlink queue of the end device with the given downlink queues updater.
+// This method rollbacks any changes to the sessions and resets the queue if the replacement fails.
+func (as *ApplicationServer) changeDownlinkQueue(ctx context.Context, dev *ttnpb.EndDevice, link *link, f downlinkQueuesUpdater) (err error) {
+	logger := log.FromContext(ctx)
+	pendingSession := dev.PendingSession
+	if pendingSession != nil {
+		logger = logger.WithField("pending_session_key_id", pendingSession.SessionKeyID)
+	}
+	currentSession := dev.Session
+	if currentSession != nil {
+		logger = logger.WithField("current_session_key_id", currentSession.SessionKeyID)
```

Comment:
Hmm not sure if we need to call this current. Do we do this elsewhere in NS or AS?

Afaik we have "pending session" and "session". If we need to clarify anything about the latter, it would be that it's the session confirmed by the end device.